doc/design.md

```diff
@@ -0,0 +1,168 @@
+
+Since remotely querying the *history* and *depth* QoS policies is not possible in many implementations (e.g. it is not required by the DDS spec),
+the bridge should offer a way to configure these values explicitly.
+This is important so that users can configure the bridge to avoid exceeding the history queue depth, resulting in lost messages.
+
+6. Remapping
+
+[Remapping](https://design.ros2.org/articles/static_remapping.html) is a generally useful concept in ROS.
+The bridge should support remapping topic, service, and action names when mapping the name from one domain to another.
+It is possible to bridge different topic names with the same name from different domains, so we should keep this in mind when it comes to remapping names.
+
+For example, users may want to remap topic `/foo` in domain ID `1` to a different name `/bar` in domain ID `2`.
+
+7. Easily configurable
+
+Since the set of bridged data depends on the application, we need an easy way to configure what data is bridged.
+Users should be able to specify the following information at an API level:
+
+- The domain IDs that should be bridged
+- The topic names and types that should be bridged (same for services and actions)
+- History and depth QoS policies for each topic bridged
+- Remappings of topic, service, and action names
+- Optionally, it could be useful to override QoS settings for each entity being bridged.
+
+Not only should the bridge be configurable at an API level, but there should also be a way to configure it without having to change code (e.g. a passed in YAML or XML file).
+
+## Approach
+
+The proposed solution involves several nodes in a single bridge process.
+For each unique domain ID involved in the bridge process, a ROS context is created for that domain ID and a node is associated with the context.
+Therefore, we end up with one node per context per domain ID, all inside a single process.
+Bridged data will be shared between the nodes in the process so that it can be rebroadcast to a different domain.
+Nodes will be created as needed, based on the configuration of the bridge.
+
+For example, the following diagram illustrates a configuration where the "/chatter" topic from domain IDs `1` and `3` are both being bridged to domain ID `2`:
+
+![](topic_example.png)
+
+### API
+
+A C++ library with a public API is provided for users to call in their own process and extend as they like.
+The public API is expected to evolve over time, but at the very least users should be able to bridge ROS networks primitives
+
+For convenience, a standalone binary executable is also provided for easy integration into ROS systems.
+
+### Supporting generic types
+
+When bridging data, serialized data will be read from one domain and forwarded to another domain.
+As long as the bridge does not need access to the typed data, there is no need for deserialization and, therefore, C++ type support.
```

Comment:
Good question! I hadn't thought about how ROS 2 security might play into the domain bridge. AFAIK, ROS 2 security is configured by specifying rules for which node can communicate over what topics (and in what direction). I *think* this means that if someone wants to enable security with a domain bridge in their system, they would need to account for the nodes being created by the bridge itself, otherwise the bridge simply won't work. I'm also not sure how security deals with different domains, I expect one might have to define separate security policies for each domain ID. It's likely if security is *not* enabled for one half of the bridge, you are exposing the bridged data.

I'll do some investigation and add a section to the design doc about security.

Reply:
I've started a section on security in the design doc (253aeb9). I *think* security will "just work", but we may need to document some best practices for enabling it. I'll come back and fill in the TODO in a follow-up PR once we've worked out a solution and document it.

Reply:
Great. Good to document the assumptions and implications as we understand them.